File: languages/csharp/projects/03-advanced/solutions/main.cs
using System;

sealed class Course
{
    private readonly string title;
    private readonly int capacity;
    private int enrolled;

    public Course(string titleValue, int capacityValue)
    {
        title = titleValue;
        capacity = capacityValue < 0 ? 0 : capacityValue;
        enrolled = 0;
    }

    public bool Enroll()
    {
        if (enrolled >= capacity)
        {
            return false;
        }

        enrolled++;
        return true;
    }

    public bool Drop()
    {
        if (enrolled <= 0)
        {
            return false;
        }

        enrolled--;
        return true;
    }

    public void PrintStatus()
    {
        Console.WriteLine($"Course: {title} | {enrolled}/{capacity} enrolled");
    }
}

class Program
{
    static void Main()
    {
        Course csharpBasics = new Course("CSharpBasics", 2);
        Course algorithms = new Course("Algorithms", 3);

        csharpBasics.Enroll();
        csharpBasics.Enroll();
        csharpBasics.Enroll();

        algorithms.Enroll();

        csharpBasics.PrintStatus();
        algorithms.PrintStatus();
    }
}

File: languages/cpp/projects/03-advanced/solutions/main.cpp
#include <iostream>
#include <string>

class Course {
  public:
    Course(const std::string& titleValue, int capacityValue)
        : title(titleValue), capacity(capacityValue), enrolled(0) {
        if (capacity < 0) {
            capacity = 0;
        }
    }

    bool enroll() {
        if (enrolled >= capacity) {
            return false;
        }
        ++enrolled;
        return true;
    }

    bool drop() {
        if (enrolled <= 0) {
            return false;
        }
        --enrolled;
        return true;
    }

    void printStatus() const {
        std::cout << "Course: " << title << " | " << enrolled << "/" << capacity << " enrolled\n";
    }

  private:
    std::string title;
    int capacity;
    int enrolled;
};

int main() {
    Course cppBasics("CppBasics", 2);
    Course algorithms("Algorithms", 3);

    cppBasics.enroll();
    cppBasics.enroll();
    cppBasics.enroll();

    algorithms.enroll();

    cppBasics.printStatus();
    algorithms.printStatus();

    return 0;
}
